### ISSUE TITLE
Pattern matching on arrays/strings

### ISSUE THREAD

**Issue by wavebeem** [squiggle-lang/squiggle-lang#99]
currently the `[a, b]` and `[a, b, ...cs]` pattern matching forms work on strings. The only intention with regard to not checking `Array.isArray` is to allow for array-like objects such as arguments and DOM Collections to work seamlessly. The behavior of `[a, b, ...cs]` on a string is pretty weird right now, too:

```
match "foobar"
case [a, b, ...cs] then
  [a, b, cs]
end
#=> ["f", "o", ["o", "b", "a", "r"]]
```

I don't think this should be encouraged as a way to process text, and throwing an error here could help catch errors sooner. In particular, this text process technique will split astral plane characters like emojis.

In short, pattern matching for arrays should check to make sure they're not strings.

Array pattern matching needs to check `Array.isArray` on the argument too because it currently throws on null/undefined.

**Comment by wavebeem** [comment]
I might fix this by just requiring array destructuring to operate on real arrays, not just array-like objects. It's not too much work to just manually convert them.
